Implement String#undump

tadd opened the pull request:
This PR is for additional information about https://bugs.ruby-lang.org/issues/12275.

tadd commented on string.c:
```diff
@@ -3418,13 +3419,34 @@ str_casecmp_p(VALUE str1, VALUE str2)
 }
 
+static long
+strseq_core(const char *str_ptr, const char *str_ptr_end, long str_len,
```
Yes, this function is extracted from `rb_strseq_index` because I needed to use this from `undump_after_backslash` too.

shyouhei commented on string.c:
```diff
@@ -6069,6 +6077,259 @@ rb_str_dump(VALUE str)
+    source_format = check_undump_source_format(s, s_end, len, enc,
+					       &forced_enc_str, &forced_enc_str_len);
+    if (source_format == UNDUMP_SOURCE_INVALID) {
+	rb_raise(rb_eArgError, "not wrapped with '\"' nor '\"...\".force_encoding(\"...\")' form");
```
Should this be ArgumentError? It seems this method has no argument.

tadd replied:
Thank you for comment.

> It seems this method has no argument.

Indeed.

But I couldn't find suitable one from built-in exceptions...
https://docs.ruby-lang.org/en/trunk/Exception.html

I choosed `RuntimeError` in 8db4fc678a, but better proposal is welcome.

nurse commented:
This failes compile with clang
```
string.c:6215:12: error: implicit conversion loses integer precision: 'unsigned long' to
      'unsigned int' [-Werror,-Wshorten-64-to-32]
            hex = ruby_scan_hex(hexstr, pos, &hexlen);
                ~ ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
string.c:6231:25: error: implicit conversion loses integer precision: 'unsigned long' to
      'unsigned int' [-Werror,-Wshorten-64-to-32]
            unsigned int hex = ruby_scan_hex(s+1, 4, &hexlen);
                         ~~~   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
string.c:6245:7: error: implicit conversion loses integer precision: 'unsigned long' to
      'unsigned int' [-Werror,-Wshorten-64-to-32]
        c2 = ruby_scan_hex(s+1, 2, &hexlen);
           ~ ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
3 errors generated.
```

nurse commented on string.c:
```diff
@@ -6069,6 +6077,259 @@ rb_str_dump(VALUE str)
+	    break;
+	}
+	n2 = rb_enc_mbclen(s+1, s_end, *penc);
+	if (n2 == 1 && IS_EVSTR(s+1, s_end)) {
```
This logic seems not needed.

tadd replied:
> This logic seems not needed.

Thanks, I may have been a bit confused.  I removed it and added some tests in c6a58b01f1.

tadd commented:
I'll rebase and resolve conflict after fixing the points that pointed out by nurse.

tadd commented:
nurse, thanks for your accurate comments.

> This failes compile with clang

I confirmed failure with my local clang-3.9.  I fixed this with 55f066bb3f.

> dumped string contains only US-ASCII strings. Therefore it is always single byte optimizable. 

Hmmm... that's true.  I'll try optimizing in next year...

BTW, I added non-ASCII character check in 1f0d0d8838.  I think this check is needed for 

* to make the specification clear to the user
* to leave room for optimization

Please let me know if this is overkill.

> current implementation doesn't check double quotes for example %[""""].undump

Oops, good catch.  I added the check in 0df3dabe1.

> more tests about escapes and invalid strings is required

Actually, I also wanted to do so.  To be honest, I was trying to match the custom of
test/ruby/test_*.rb without thinking deeply.

I added more tests in 0cdf95bcd.

tadd commented:
OK, I'll try to rebase and resolve conflict now...

tadd commented:
I added 3 commits after resolving conflict:

* 670ad3a: unescaped character expression is adjusted as same as parse.y.
    * This was advised by @mame. Thanks!
* ee2059c: reject input string if it contained NUL `\0` byte. (`dump`-ed string won't contain that)
    * This is a continuation of 1f0d0d8.
    * I believe "narrow entrance" is better but might be overkill..?
* 5091b01174f: needless (duplicated) `const` keyword removed

Then, I'll begin to create v3.patch to attach to the issue page.
